### PR TITLE
[WIP] Refactor openmp

### PR DESCRIPTION
**Pull request by cz4rs**
This is a temporary test PR before submitting to upstream repo.

### CODE REVIEW

**Comment by cz4rs** [comment]
sample benchmark results:
<details>
<summary>refactor-openmp</summary>
<p>

```
kokkos  build_benchmarks  ../benchmarks/benchmark_suite/scripts/run_benchmark.bash ../benchmarks/benchmark_suite/scripts/ OpenMP SNB mpicxx
BytesAndFlops: 0
0
0
0
0.125739 0
0
0
0
0.046369
MiniMD: 0.198257 0.261246
MiniFE: 0.325626 0.523059
Total Result:  0 0 0 0 0.0209565 0 0 0 0.22576

kokkos  build_benchmarks  ../benchmarks/benchmark_suite/scripts/run_benchmark.bash ../benchmarks/benchmark_suite/scripts/ OpenMP SNB mpicxx
BytesAndFlops: 0
0
0
0
0.125619 0
0
0
0
0.0479299
MiniMD: 0.202339 0.321094
MiniFE: 0.320804 0.483669
Total Result:  0 0 0 0 0.0209365 0 0 0 0.229306
```

</p>
</details>  

<details>
<summary>develop</summary>
<p>

```
kokkos  build_benchmarks  ../benchmarks/benchmark_suite/scripts/run_benchmark.bash ../benchmarks/benchmark_suite/scripts/ OpenMP SNB mpicxx
BytesAndFlops: 0
0
0
0
0.122329 0
0
0
0
0.0462413
MiniMD: 0.199624 0.319641
MiniFE: 0.322931 0.519649
Total Result:  0 0 0 0 0.0203882 0 0 0 0.234681
```

</p>
</details>

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp [hunk 1]
```diff
@@ -103,43 +103,99 @@ class ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::OpenMP> {
...
+    auto ibeg = m_policy.begin();
+    auto iend = m_policy.end();
+
+#pragma omp parallel for schedule(dynamic, m_policy.chunk_size()) \
+    num_threads(OpenMP::impl_thread_pool_size())
+#ifdef KOKKOS_ENABLE_AGGRESSIVE_VECTORIZATION
+#ifdef KOKKOS_ENABLE_PRAGMA_IVDEP
+#pragma ivdep
+#endif
+#endif
```
Comment (PhilMiller) [inline core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:134]: Using the alternative syntax `_Pragma(ivdep)`, you can then define a macro that expands to either that or nothing, so we don't need the huge ugly `#if` at each usage site

Comment (PhilMiller) [reply core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp:134]: e.g. `#define KOKKOS_PRAGMA_IVDEP_IF_ENABLED _Pragma(ivdep)`

**Comment by PhilMiller** [comment]
This looks pretty straightforwardly good to me.

File: core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp (hunk 1 above)
Comment (cz4rs) [reply core/src/OpenMP/Kokkos_OpenMP_Parallel.hpp]: Done (f81588f).

**Comment by cz4rs** [comment]
Draft PR with refactored `Kokkos::parallel_for` has been submitted to upstream repo: https://github.com/kokkos/kokkos/pull/4664.

**Comment by cz4rs** [comment]
Replaced with https://github.com/kokkos/kokkos/pull/4664.

**Pull request closed without merging**